nifi-nar-bundles/nifi-neo4j-bundle/nifi-neo4j-processors/src/main/java/org/apache/nifi/processors/neo4j/AbstractNeo4JCypherExecutor.java

```diff
@@ -0,0 +1,281 @@
+            .defaultValue("100")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor MAX_CONNECTION_ACQUISITION_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("neo4j-max-connection-acquisition-timeout")
+            .displayName("Neo4J Max Connection Acquisition Timeout")
+            .description("The maximum connection acquisition timeout.")
+            .defaultValue("60 second")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor IDLE_TIME_BEFORE_CONNECTION_TEST = new PropertyDescriptor.Builder()
+            .name("neo4j-idle-time-before-test")
+            .displayName("Neo4J Idle Time Before Connection Test")
+            .description("The idle time before connection test.")
+            .defaultValue("60 seconds")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor MAX_CONNECTION_LIFETIME = new PropertyDescriptor.Builder()
+            .name("neo4j-max-connection-lifetime")
+            .displayName("Neo4J Max Connection Lifetime")
+            .description("The maximum connection lifetime")
+            .defaultValue("3600 seconds")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor ENCRYPTION = new PropertyDescriptor.Builder()
+            .name("neo4j-encryption")
+            .displayName("Neo4J Encrytion")
+            .description("Is connection encrypted")
+            .defaultValue("true")
+            .required(true)
+            .allowableValues("true","false")
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static AllowableValue TRUST_SYSTEM_CA_SIGNED_CERTIFICATES =
+        new AllowableValue(TrustStrategy.Strategy.TRUST_SYSTEM_CA_SIGNED_CERTIFICATES.name(),
+            "Trust System CA Signed Certificates", "Trust system specified CA signed certificates");
+
+    public static AllowableValue TRUST_CUSTOM_CA_SIGNED_CERTIFICATES =
+        new AllowableValue(TrustStrategy.Strategy.TRUST_CUSTOM_CA_SIGNED_CERTIFICATES.name(),
+            "Trust Custom CA Signed Certificates", "Trust custom CA signed certificates defined in the file");
+
+    public static AllowableValue TRUST_ALL_CERTIFICATES =
+        new AllowableValue(TrustStrategy.Strategy.TRUST_ALL_CERTIFICATES.name(),
+            "Trust All Certificates", "Trust all certificate");
+
+    protected static final PropertyDescriptor TRUST_STRATEGY = new PropertyDescriptor.Builder()
```

Comment:
The way we have solved this problem in other processors is to allow for a *truststore* or `SSLContextService` to be provided. Please see `InvokeHTTP` for an example. If the remote Neo4J endpoint can expect NiFi to present a certificate as well, the `SSLContextService` option is definitely preferred.

Reply:
@alopresto - Neo4j driver configuration uses these strategies.  Let me know what is your recommendation of handling this.

Reply:
I think your best bet would be to support only the Custom CA option and use the `SSLContextService` to build up the configuration for that. @alopresto does that seem right to you?

Reply:
I believe that Neo4J has abstracted/encapsulated it's security implementation and exposed them as trust strategies to be set on the configuration builder.  I think we should use these strategies/interfaces so that our code is insulated from any internal changes/optimizations in the neo4j driver. 
Let me know what are your thoughts/recommendations.

Reply:
@mans2singh something else to keep in mind is that many NiFi users may not be especially security savvy so steering them toward safe options and hiding dangerous ones can be perfectly reasonable given that you're targeting an audience of varying levels of knowledge.

Reply:
@alopresto @MikeThomsen  - I wanted to understand your recommendations - 
1. Remove the trust all strategy.
2. Use SSLContextService to configure values for the other two neo4j trust strategies to be consistent with NIFI ssl configuration pattern.
I will explore these options and please let me know if I've missed anything.  
Mans

Reply:
(Caveat: Final decision is with @alopresto if he disagrees because he's the resident crypto expert on the team)

1. I think trusting all certificates is a really dangerous and likely to end up aiding people in shooting their own feet off with respect to SSL here. NiFi provides a toolkit for generating certificates for users that don't have their own CA set up, so there is no reason for people to be creating their own self-signed certs and trying to force them to work (I've worked with plenty of folks that go this route; it really is an issue to think about).

2. You can also look at the Mongo and ElasticSearch examples as well. Both of them do a pretty good job of hiding the specific configuration issues for their respective endpoints behind the `SSLContextService`. The end goal is something like that; the user drops in the `SSLContextService` and you take it from there for them.

Reply:
You might also wanted to reference the NiFi TLS Toolkit in the documentation. There is an issue of "culture" with some of the platforms that NiFi supports. Some of them are more likely to have different tendencies towards things like security. For example, Mongo tends to be a little more... "wild west" when it comes to making good decisions than SQL or Hadoop users.

So not knowing anything about the culture and tendencies of Neo4J users, I would suggest pointing them to our TLS Toolkit as an option if they don't have access to an enterprise certificate authority.

Reply:
IMO this is another good reason to abstract the Neo4J stuff into a controller service implementation. The processor doesn't need to know about how to secure a connection in order to submit a Cypher query. Then the "maintenance surface" is smaller, especially if we decide to go with the Neo4J-specific security configuration.

Reply:
@MikeThomsen @mattyb149 - I am working on using the ssl context service to get the custom cert file name or default to system certifcates.  I will remove the trust all strategy.  Let me know if you have any advice.